account/account.go

```diff
@@ -75,23 +75,25 @@ type Account interface {
 type MutableAccount interface {
 	Account
 	// Set public key (needed for lazy initialisation), should also set the dependent address
-	SetPublicKey(pubKey PublicKey) MutableAccount
+	SetPublicKey(pubKey PublicKey)
 	// Subtract amount from account balance (will panic if amount is greater than balance)
-	SubtractFromBalance(amount uint64) (MutableAccount, error)
+	SubtractFromBalance(amount uint64) error
 	// Add amount to balance (will panic if amount plus balance is a uint64 overflow)
-	AddToBalance(amount uint64) (MutableAccount, error)
+	AddToBalance(amount uint64) error
 	// Set EVM byte code associated with account
-	SetCode(code []byte) MutableAccount
+	SetCode(code []byte)
 	// Increment Sequence number by 1 (capturing the current Sequence number as the index for any pending mutations)
-	IncSequence() MutableAccount
```

Comment:
This enables a builder style syntax. Obviously a matter of taste, but better not to make inconsequential changes like this.

Reply:
No need to return the modified struct (here `MutableAccount` object) when the methods are called by reference (pointer).
 
First of all, it increases the performance; because no need to copy the whole object any time a modification is required. It just changes the internal object. It makes object more encapsulated and robust. 
Also it has better code understanding, `MutableAccount` is the instance of account that can be modified. In the other hand the, the `Account` is just a read-only instance and no modifications is allowed. Any object which needs to modify Account should have `MutableAccount` instance. This policy make the code more safe.
By just returning the error code we can check if the operation have failed or not. No need to have copy of the object again.

Here is the rules that I am following to check if a method should be called by pointer or not:
* if a method doesn't change any struct field (`const` method), it calls without reference (pointer)
* if a method is not a `const` method (it changes the struct fields), it calls by reference (pointer).

```
func (s Struct)ImmutableMethod() err {
...
}

func (s* Struct)MutableMethod() err {
...
}
```

Reply:
> No need to return the modified struct (here MutableAccount object) when the methods are called by reference (pointer).

There is no strict 'need' but it allows a declarative builder syntax like:

```go
acc.IncSequence().AddToBalance(inputAmount).SubtractToBalance(outputAmount)
```
This is why the method returns the bash-in-place object. Now whether you like this builder syntax is a matter of taste. I happen to think it keeps some important operations on a single logically line and prevents accidentally interspersing operations that should be atomic with other calls (that may err for example).

More to the point though really is there is no reason to change this relating to your changes - so when we come to PR as a hope we can soon it either belongs in a separate PR (I know you didn't open this as PR of course :)! ) or should be reverted.

> First of all, it increases the performance; because no need to copy the whole object any time a modification is required. 

This is incorrect. It does not copy the object the return type is an interface which is a pointer type. The performance implications are therefore utterly negligible.

> It just changes the internal object. It makes object more encapsulated and robust.

All these methods just change the internal object. You'll have to explain why this makes them more encapsulated and robust - I don't see it.

> Also it has better code understanding, MutableAccount is the instance of account that can be modified. In the other hand the, the Account is just a read-only instance and no modifications is allowed. Any object which needs to modify Account should have MutableAccount instance. This policy make the code more safe.

This comment applies in its entirety to the current implementation - so I agree!

> By just returning the error code we can check if the operation have failed or not. No need to have copy of the object again.

You can still check the error code in the current implementation. As I have said there is not object copy.